source/validate_cfg.cpp

```diff
@@ -167,6 +167,153 @@ string ConstructErrorString(const Construct& construct,
+    uint32_t* case_fall_through, const BasicBlock* merge,
+    const std::unordered_set<uint32_t>& case_targets) {
+  std::vector<const BasicBlock*> stack;
+  stack.push_back(target_block);
+  std::unordered_set<const BasicBlock*> visited;
+  while (!stack.empty()) {
+    const auto block = stack.back();
+    stack.pop_back();
+
+    if (block == merge) continue;
+
+    if (!visited.insert(block).second) continue;
+
+    if (target_block->reachable() && block->reachable() &&
+        target_block->dominates(*block)) {
+      // Still in the case construct.
+      for (auto successor : *block->successors()) {
+        stack.push_back(successor);
+      }
+    } else {
+      // Exiting to non-merge block
+      if (!case_targets.count(block->id())) {
+        continue;
+      }
+
+      if (*case_fall_through == 0u) {
+        *case_fall_through = block->id();
+      } else if (*case_fall_through != block->id()) {
+        return _.diag(SPV_ERROR_INVALID_CFG)
+               << "Case construct headed by " << _.getIdName(target_block->id())
+               << " branches to multiple case constructs "
+               << _.getIdName(*case_fall_through) << " and "
+               << _.getIdName(block->id());
+      }
+    }
+  }
+
+  return SPV_SUCCESS;
+}
+
+spv_result_t StructuredSwitchChecks(const ValidationState_t& _,
+                                    const Function& function,
+                                    const Instruction* switch_inst,
+                                    const BasicBlock* header,
+                                    const BasicBlock* merge) {
+  std::unordered_set<uint32_t> case_targets;
+  for (uint32_t i = 1; i < switch_inst->operands().size(); i += 2) {
+    uint32_t target = switch_inst->GetOperandAs<uint32_t>(i);
+    if (target != merge->id()) case_targets.insert(target);
+  }
+  // Tracks how many times each case construct is targeted by another case
+  // construct.
+  std::map<uint32_t, uint32_t> num_fall_through_targeted;
+  // uint32_t num_targeting_default = 0u;
+  uint32_t default_case_fall_through = 0u;
+  uint32_t default_target = switch_inst->GetOperandAs<uint32_t>(1u);
+  std::unordered_set<uint32_t> seen;
+  for (uint32_t i = 1; i < switch_inst->operands().size(); i += 2) {
+    uint32_t target = switch_inst->GetOperandAs<uint32_t>(i);
+    if (target == merge->id()) continue;
+
+    if (!seen.insert(target).second) continue;
+
+    // Check each case construct for the following:
+    // 1. It is dominated by the OpSwitch.
+    // 2. It branches to at most one other case construct.
+    // 3. If it branches to another case construct, it must immediately
+    // precede that target (or the fall through of the default case) in the
+    // operand list.
+    bool is_default = i == 1;
+    const auto target_block = function.GetBlock(target).first;
+    if (header->reachable() && target_block->reachable() &&
+        !header->dominates(*target_block)) {
```

Comment:
No, it is not the same. This check is ensuring that there is no way to enter the construct without going through the header. Assuming there were no other checks, it would catch things like:
```
%1 = OpLabel
OpBranchConditional %cond %2 %3
%3 = OpLabel
OpSelectionMerge %4 None
OpSwitch %int0 %3
%3 = OpLabel
OpReturn
%4 = OpLabel
OpUnreachable
```